File: components/gitpod-protocol/src/analytics.ts
/**
 * Copyright (c) 2021 Gitpod GmbH. All rights reserved.
 * Licensed under the GNU Affero General Public License (AGPL).
 * See License-AGPL.txt in the project root for license information.
 */

export const IAnalyticsWriter = Symbol('IAnalyticsWriter');

type Identity =
    | { userId: string | number; anonymousId?: string | number }
    | { userId?: string | number; anonymousId: string | number };

interface Message {
    messageId?: string;
}

export type IdentifyMessage = Message &
    Identity & {
        traits?: any;
        timestamp?: Date;
        context?: any;
    };

export type TrackMessage = Message &
    Identity & {
        event: string;
        properties?: any;
        timestamp?: Date;
        context?: any;
    };

export type PageMessage = Message &
    Identity & {
        properties?: any;
        timestamp?: Date;
        context?: any;
    };

export type RemoteTrackMessage = Omit<TrackMessage, 'timestamp' | 'userId'>;
export type RemotePageMessage = Omit<PageMessage, 'timestamp' | 'userId'> & {
    includePII?: boolean;
};

export type RemoteIdentifyMessage = Omit<IdentifyMessage, 'timestamp' | 'userId'>;

export interface IAnalyticsWriter {
    identify(msg: IdentifyMessage): void;

    track(msg: TrackMessage): void;

    page(msg: PageMessage): void;
}

File: components/ws-proxy/pkg/analytics/analytics.go
// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
// Licensed under the GNU Affero General Public License (AGPL).
// See License.AGPL.txt in the project root for license information.

package analytics

import (
	"github.com/gitpod-io/gitpod/common-go/analytics"
)

var writer analytics.Writer = analytics.NewFromEnvironment()

func Identify(message analytics.IdentifyMessage) {
	writer.Identify(message)
}
func Track(message analytics.TrackMessage) {
	writer.Track(message)
}
func Close() error {
	return writer.Close()
}
